GpuTimer: improve kernel execution time measurement accuracy

olupton opened the pull request:
This PR changes how `GpuTimer` measures execution time, which should improve the measurement accuracy.
Quoting my comment in the code:
```c++
// When a timer is created it launches a delay kernel into the given stream and
// queues a start event immediately afterwards. This delay kernel blocks
// execution on the stream until GetElapsedDuration() is called, at which point
// an end event is queued and the delay kernel exits. This allows the device
// execution time of the tasks queued to the stream while the timer is active
// to be measured more accurately.
```
this should improve the accuracy of the measurements that are used to make auto-tuning decisions, especially for small kernels.

There are a couple of edge cases that have special treatment:
- if `CUDA_LAUNCH_BLOCKING=1` then the delay kernel will not achieve anything, so we don't launch it
- if any code in the timed region synchronises the device then the host will wait for the delay kernel to time out before actually launching the kernels to be measured. This will lead to a poor quality measurement, and an error message is printed. This condition can be met if one of the kernels being measured is lazily loaded, as lazy loading can trigger synchronisation. Best practice is to execute a warmup run (without timing enabled) before the timed execution.

Two parts of the autotuning code are updated to follow this best practice.

The `GpuTimer::Create*` signatures that were deprecated in https://github.com/openxla/xla/pull/9841 do not benefit from these accuracy improvements.

cc: @sergachev @nouiz

thomasjoerg commented on xla/stream_executor/gpu/gpu_timer_kernel.cu.cc:
```diff
@@ -0,0 +1,46 @@
+// Wait for the value pointed to by `semaphore` to have value `target`, timing
+// out after approximately `APPROX_TIMEOUT_SECONDS` seconds if that value is
+// not reached. This can happen if, for example, blocking launches are enabled
+// via CUDA_LAUNCH_BLOCKING=1.
```
Will it be feasible to use `CUDA_LAUNCH_BLOCKING=1` on XLA GPU after this change. A timeout of 10s may increase compile time by a lot.

olupton replied:
That variable is [only for debugging](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#concurrent-execution-between-host-and-device), and I hope it's rarely used. That said, we could probably make the timeout shorter without affecting the auto-tuning results in normal execution. In any case, this change would only slow down compile time, it shouldn't affect correctness.

If we could detect that blocking launches are enabled, we could just skip the extra kernel as it's not going to achieve anything in that case.

thomasjoerg replied:
True, but compile time matters for debugging. It could at least be very surprising to have compile time affected by `CUDA_LAUNCH_BLOCKING=1` potentially much.

> If we could detect that blocking launches are enabled, we could just skip the extra kernel as it's not going to achieve anything in that case.

Right, that's what I was thinking. Not sure what the best way would be. I asked Ilia and Ferderik for a review. Maybe they have ideas :-)

nouiz replied:
I don't like this idea. This will mess with the autotune decision and so the optimized graph will be different. So we will hide a significant category of issue.
Why 10s is too much. 1s is clearly better. But we can probably go even smaller. 
We don't compile and don't include the warm-up in the timing.
@olupton can you look how long it takes one of our lib to dispatch? Maybe 10x that time? 
Why not go down to 100ms?

nouiz replied:
Olli made me realize, with CUDA_LAUNCH_BLOCKING=1, the delay kernel won't do what it should.
So it is useless to have it and we can't try to get close results to the normal path.
I would still lower the timeout to 1s, just in case. I don't think doing lower will be useful as it shouldn't be triggered.

olupton replied:
OK. I will:
- bypass the delay kernel when `CUDA_LAUNCH_BLOCKING` is set to `1`.
- decrease the timeout to a nominal 1 second
- push a fix to the CI failure (which is the result of a _different_ way of hitting the timeout)

cheshire commented:
Splitting the refactoring from a functional change sounds great either way.

olupton commented:
> Splitting the refactoring from a functional change sounds great either way.

Done in https://github.com/openxla/xla/pull/9841.

sergachev commented on xla/service/gpu/conv_algorithm_picker.cc:
```diff
@@ -622,15 +621,33 @@ absl::StatusOr<AutotuneResult> GpuConvAlgorithmPicker::AutotuneOneConvRunner(
-  for (;
-       num_iters < kMaxIter && launch_status.ok() && profile_result.is_valid();
-       num_iters++) {
+  for (; num_iters < kMaxIter; num_iters++) {
```
++num_iters (https://google.github.io/styleguide/cppguide.html#Preincrement_and_Predecrement)

olupton replied:
Done in 3dcc5caf0a.

sergachev commented on xla/stream_executor/gpu/gpu_timer_kernel.h:
```diff
@@ -0,0 +1,17 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+namespace stream_executor::gpu {
```
Doesn't this header need the usual #ifndef guard?

olupton replied:
Right, good catch! Done in 3dcc5caf0a.

sergachev commented on xla/stream_executor/gpu/gpu_timer.cc:
```diff
@@ -51,10 +51,21 @@ absl::Duration RandomDuration() {
   return absl::Microseconds(distribution(rng));
 }
 
+bool LaunchDelayKernel() {
```
The name is a little misleading because the function does not actually launch anything. Maybe ShouldLaunchDelayKernel()?

olupton replied:
Changed, thanks.

sergachev commented on xla/stream_executor/gpu/gpu_timer.cc:
```diff
@@ -72,23 +83,90 @@ absl::Duration RandomDuration() {
+GpuTimer::GpuSemaphore::Create(StreamExecutor* executor) {
+  // Check the assumption that this device supports unified addressing.
+  GpuExecutor* gpu_executor = ExtractGpuExecutor(executor);
+  TF_ASSIGN_OR_RETURN(int status, (GpuDriver::GetDeviceAttribute(
```
Are the parentheses around GpuDriver::GetDeviceAttribute() necessary?

olupton replied:
No, I think that was just reflexive... Removed.

sergachev commented:
LGTM

nouiz reviewed (approved):
LGTM

nouiz commented on xla/stream_executor/gpu/gpu_timer.cc:
```diff
@@ -51,10 +51,21 @@ absl::Duration RandomDuration() {
+  // Only launch the delay kernel if CUDA_LAUNCH_BLOCKING is not set to 1.
+  static bool value = [] {
+    const char* blocking = std::getenv("CUDA_LAUNCH_BLOCKING");
+    return !(blocking && std::string_view{blocking} == "1");
```
What happens values others then 0 and 1? Should this be != "0"?

nouiz replied:
Should this also check for the Unified memory?
If there isn't any unified memory, I think we should disable the delay kernel instead of crashing.
But I don't know if the supported GPU all support it or not. So maybe this isn't useful.

olupton replied:
https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#cuda-environment-variables documents that the only valid values are 0 and 1. From a quick test, it seems like the value really has to be 1 to enable blocking launches, other truth-y values don't work. Given that, I think the code is correct as-is.

I think you mean unified addressing, not unified (aka managed) memory? I think it's probably not worth worrying about, from https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#unified-virtual-address-space it looks like the check should succeed on any 64bit system with a compute capability >=2.0 GPU.

olupton replied:
Looks like 10 years ago, the K1. I imagine this will never be an issue...but I'll push a change to skip the delay kernel if support isn't detected.

olupton replied:
Done in be8d70fa80.

olupton commented:
Rebased on main. IIUC this needs attention from a Googler to apply the changes to TSL.

olupton commented:
@thomasjoerg I rebased and fixed the GPU test failures, sorry for the delay.
The ARM64 failure looks to be red on all recent PRs, and I didn't see any sign of it being linked to this one, so I'm ignoring it.

cheshire commented:
@olupton Could you update the commit message with data on precision difference in autotuning before/after?

olupton commented:
> @olupton Could you update the commit message with data on precision difference in autotuning before/after?

Done, I squashed the commits and expanded the description in the remaining commit message.

cheshire reviewed (changes requested):
Question: this modifies cuDNN and cuBLAS autotuning, but not Triton. Does it not need modifications?

Another question: is there a noticeable effect on autotuning time? Could we have some before/after measurements?

olupton commented:
> Question: this modifies cuDNN and cuBLAS autotuning, but not Triton. Does it not need modifications?
> 
> Another question: is there a noticeable effect on autotuning time? Could we have some before/after measurements?

The cuDNN and cuBLAS autotuners are modified because they didn't follow the new best practice of having a warmup iteration **that is not timed**. The Triton autotuner follows a different code path that was already doing the right thing.

I have not noticed any effect, but I didn't try carefully to put a number on that. Note that there is only one extra autotuned kernel execution in one path (`gemm_algorithm_picker.cc`), otherwise this is just shuffling how/when the events are recorded with the delay kernel (which is itself very fast). I don't see a reason to expect a relevant difference.

In the ~30µs runtime example described in the commit message, the wallclock time of autotuning without this PR was 250ms and with this PR it was 224ms. That's just a single measurement, and the difference is probably just noise, but at least it's a data point showing nothing is blowing up, as expected.

olupton commented on xla/stream_executor/gpu/gpu_timer.cc:
```diff
@@ -51,10 +51,21 @@ absl::Duration RandomDuration() {
+  // Only launch the delay kernel if CUDA_LAUNCH_BLOCKING is not set to 1.
+  static bool value = [] {
+    const char* blocking = std::getenv("CUDA_LAUNCH_BLOCKING");
+    return !(blocking && std::string_view{blocking} == "1");
```
Changed (to my eyes, they're as readable as each other 🤷)

dimitar-asenov commented on xla/stream_executor/gpu/gpu_timer_kernel.h:
```diff
@@ -0,0 +1,26 @@
+}  // namespace delay_kernel
+}  // namespace stream_executor::gpu
+
+#endif
```
Please add `  // TENSORFLOW_COMPILER_XLA_STREAM_EXECUTOR_GPU_GPU_TIMER_KERNEL_H_`

(note the 2 spaces between `#endif` and the comment.

dimitar-asenov replied:
Actually the comment should be `//XLA_STREAM_EXECUTOR_GPU_GPU_TIMER_KERNEL_H_`

olupton replied:
Done

dimitar-asenov commented on xla/stream_executor/gpu/BUILD:
```diff
@@ -350,7 +357,9 @@ gpu_only_cc_library(
         "@tsl//tsl/platform:errors",
         "@tsl//tsl/platform:status",
         "@tsl//tsl/platform:statusor",
```
Please add `":gpu_timer_kernel_header",` to the unconditional list of deps.

dimitar-asenov reviewed (approved):
I left a couple of minor comments. Please also rebase on the latest code, as there as some conflicting changes.

reedwm commented:
Sorry, I had to roll this back since it broke some internal tests. Will have more details soon.

reedwm commented:
@dimitar-asenov and @beckerhe are looking into this.

dimitar-asenov commented:
Henning has a potential fix. We can verify it tomorrow and if things work out well, we can resubmit this.
